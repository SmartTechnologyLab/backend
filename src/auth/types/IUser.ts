export interface IUser {
  displayName: string;
  email: string;
  accessToken: string;
  refreshToken: string;
  id?: string;
}
