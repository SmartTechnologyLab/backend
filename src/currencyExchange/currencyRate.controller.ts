import {
  BadRequestException,
  Controller,
  Get,
  Injectable,
  Query,
} from '@nestjs/common';
import { CurrencyRateService } from './currencyRate.service';

@Injectable()
@Controller('currency-rate')
export class CurrencyRateController {
  constructor(private currencyExchangeService: CurrencyRateService) {}

  @Get()
  async getUahCurrenyRate(
    @Query('currency') currency: string,
    @Query('date') date: string,
  ) {
    try {
      const rate = await this.currencyExchangeService.getCurrencyExchange(
        currency,
        date,
      );
      console.log(rate);
      return rate;
    } catch (error) {
      throw new BadRequestException('Error while getting currency rate');
    }
  }
}
