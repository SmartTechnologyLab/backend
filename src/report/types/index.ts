type CorporateActionTypeId = 'dividend' | 'maturity';

type OperationType = 'buy' | 'sell';

export interface ICorporateAction {
  date: string;
  type: string;
  type_id: CorporateActionTypeId;
  amount: number;
  ticker: string;
  isin: string;
  currency: string;
}

export interface IPosition {
  i: string;
  q: number;
}

export interface IDealReport<T> {
  total: number;
  totalTaxFee: number;
  totalMilitaryFee: number;
  deals: T[];
}

export interface ITrades {
  trade_id: string;
  date: string;
  p: number;
  q: number;
  commission: number;
  operation: OperationType;
  summ: number;
  instr_nm: string;
  curr_c: string;
}

export interface IReport {
  date_start: string;
  corporate_actions: {
    detailed: ICorporateAction[];
  };
  trades: {
    detailed: ITrades[];
    securities: Record<string, number>;
  };
}

export type DealOptions = {
  purchaseCommission?: number;
  purchaseDate?: Date;
  purchasePrice?: number;
  purchaseRate?: number;
  quantity?: number;
  saleCommission?: number;
  saleDate?: Date;
  salePrice?: number;
  saleRate?: number;
  ticker?: string;
};

export interface IDealShort {
  ticker: string;
  total: number;
  percent: number;
  purchaseUAH: number;
  saleUAH: number;
}

export type Deal = {
  id: symbol;
  percent: number;
  purchase: {
    date: Date;
    price: number;
    sum: number;
    commission: number;
    rate: number;
    uah: number;
  };
  quantity: number;
  sale: {
    date: Date;
    price: number;
    sum: number;
    commission: number;
    rate: number;
    uah: number;
  };
  ticker: string;
  total: number;
};
