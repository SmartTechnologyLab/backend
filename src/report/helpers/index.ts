import { sort } from 'ramda';
import { IReport } from '../types';

export const sortByDate = (report: IReport[]): IReport[] => {
  const byDate = (a: IReport, b: IReport) =>
    new Date(b.date_start).getTime() - new Date(a.date_start).getTime();
  return sort(byDate, report);
};
