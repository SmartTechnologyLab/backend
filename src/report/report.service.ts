import { BadRequestException, Injectable } from '@nestjs/common';
import { groupBy, clone } from 'ramda';
import { CurrencyExchangeService } from '../currencyExchange/currencyExchange.service';
import {
  Deal,
  DealOptions,
  ICorporateAction,
  IDealReport,
  IReport,
  ITrades,
} from './types';

@Injectable()
export class ReportService {
  constructor(private currencyExchangeService: CurrencyExchangeService) {}

  readReport(file: Express.Multer.File): IReport {
    try {
      const reportContent = file.buffer.toString('utf-8');
      return JSON.parse(reportContent);
    } catch (error) {
      throw new Error(error);
    }
  }

  groupTradesByTicker(trades: ITrades[]): { [key: string]: ITrades[] } {
    const modifiedTrades = trades?.map((deal) => {
      const ticker = deal.instr_nm.split('.').at(0);
      return { ...deal, instr_nm: ticker };
    });

    return groupBy((deal: ITrades) => deal.instr_nm, modifiedTrades);
  }

  async getReportExtended(report: ITrades[]): Promise<IDealReport<Deal>> {
    const deals: Deal[] = [];

    const groupedDeals = this.groupTradesByTicker(clone(report));
    for (const ticker in groupedDeals) {
      let buyQueue: ITrades[] = [];
      let sellComission = 0;
      for (const [index, deal] of Object.entries(groupedDeals[ticker])) {
        if (deal.operation === 'buy' && deal.q > 0) {
          const existingDeal = this.findDealByDateAndPrice(buyQueue, deal);
          if (existingDeal) {
            existingDeal.q += deal.q;
            existingDeal.commission += deal.commission;
          } else {
            buyQueue.push(deal);
          }
        } else if (deal.operation === 'sell' && buyQueue.length > 0) {
          sellComission = deal.commission / deal.q;

          for (const purchaseDeal of buyQueue) {
            if (deal.q > 0 && purchaseDeal.q > 0) {
              const newDeal = await this.setDeal(
                purchaseDeal,
                deal,
                sellComission,
              );

              deals.push(newDeal);
            }
          }

          buyQueue = buyQueue.filter((purchaseDeal) => purchaseDeal.q > 0);

          while (
            deal.q > 0 &&
            (buyQueue.some((b) => b.q > 0) ||
              groupedDeals[ticker].find(
                (d, indx) => d.operation === 'buy' && indx > +index,
              ))
          ) {
            const foundShortBuy = buyQueue.some((b) => b.q > 0)
              ? buyQueue.filter((b) => b.q > 0).at(0)
              : groupedDeals[ticker].find(
                  (d, indx) => d.operation === 'buy' && indx > +index,
                );

            if (foundShortBuy && foundShortBuy.q > 0) {
              const newDeal = await this.setDeal(
                foundShortBuy,
                deal,
                sellComission,
              );

              deals.push(newDeal);
            }
          }
        } else if (
          deal.operation === 'sell' &&
          buyQueue.length === 0 &&
          groupedDeals[ticker].find(
            (t, ind) => t.operation === 'buy' && t.q > 0 && +index < ind,
          )
        ) {
          sellComission = deal.commission / deal.q;

          const foundShortBuy = groupedDeals[ticker].find(
            (t, ind) => t.operation === 'buy' && t.q > 0 && +index < ind,
          );

          if (foundShortBuy) {
            const newDeal = await this.setDeal(
              foundShortBuy,
              deal,
              sellComission,
            );

            deals.push(newDeal);
          }
        }
        sellComission = 0;
      }

      if (
        groupedDeals[ticker].some(
          (deal) => deal.operation === 'sell' && deal.q > 0,
        )
      ) {
        throw new BadRequestException('Not enough buy deals');
      }
    }

    const total = deals.reduce((acc, deal) => acc + deal.total, 0);

    const totalTaxFee = this.getTotalTaxFee(total);

    const totalMilitaryFee = this.getMilitaryFee(total);

    return {
      total,
      totalTaxFee,
      totalMilitaryFee,
      deals: deals.sort((a, b) => a.ticker.localeCompare(b.ticker)),
    };
  }

  async getPrevTrades(report: ITrades[]): Promise<ITrades[]> {
    const remainedPurchaseDeals: ITrades[] = [];

    const groupedDeals = this.groupTradesByTicker(clone(report));

    for (const ticker in groupedDeals) {
      let buyQueue: ITrades[] = [];
      for (const [index, deal] of Object.entries(groupedDeals[ticker])) {
        if (deal.operation === 'buy' && deal.q > 0) {
          const existingDeal = this.findDealByDateAndPrice(buyQueue, deal);
          if (existingDeal) {
            existingDeal.q += deal.q;
            existingDeal.commission += deal.commission;
          } else {
            buyQueue.push(deal);
          }
        } else if (deal.operation === 'sell' && buyQueue.length > 0) {
          for (const purchaseDeal of buyQueue) {
            if (deal.q >= purchaseDeal.q) {
              const quantityToProcess = Math.min(purchaseDeal.q, deal.q);
              deal.q -= quantityToProcess;
              purchaseDeal.q -= quantityToProcess;
            }
          }

          buyQueue = buyQueue.filter((purchaseDeal) => purchaseDeal.q > 0);

          while (
            deal.q > 0 &&
            (buyQueue.some((b) => b.q > 0) ||
              groupedDeals[ticker].find(
                (d, indx) => d.operation === 'buy' && indx > +index,
              ))
          ) {
            const foundShortBuy = buyQueue.some((b) => b.q > 0)
              ? buyQueue.filter((b) => b.q > 0).at(0)
              : groupedDeals[ticker].find(
                  (d, indx) => d.operation === 'buy' && indx > +index,
                );

            if (foundShortBuy && foundShortBuy.q > 0) {
              const quantityToProcess = Math.min(foundShortBuy.q, deal.q);
              foundShortBuy.q -= quantityToProcess;
              deal.q -= quantityToProcess;
            }
          }
        } else if (
          deal.operation === 'sell' &&
          buyQueue.length === 0 &&
          groupedDeals[ticker].find(
            (t, ind) => t.operation === 'buy' && t.q > 0 && +index < ind,
          )
        ) {
          const foundShortBuy = groupedDeals[ticker].find(
            (t, ind) => t.operation === 'buy' && t.q > 0 && +index < ind,
          );

          if (foundShortBuy) {
            const quantityToProcess = Math.min(foundShortBuy.q, deal.q);
            deal.q -= quantityToProcess;
            foundShortBuy.q -= quantityToProcess;
          }
        }
      }
      remainedPurchaseDeals.push(...buyQueue.filter((deal) => deal.q > 0));
    }

    return remainedPurchaseDeals;
  }

  async getReport(report: ITrades[]): Promise<IDealReport<Deal>> {
    const trades = (await this.getReportExtended(report)) as IDealReport<Deal>;

    const deals = Object.values(
      trades.deals.reduce(
        (acc, deal) => {
          if (!acc[deal.ticker]) {
            acc[deal.ticker] = deal;
          } else {
            acc[deal.ticker].sale.uah += deal.sale.uah;
            acc[deal.ticker].purchase.uah += deal.purchase.uah;
            acc[deal.ticker].total += deal.total;
          }
          return acc;
        },
        {} as Record<string, Deal>,
      ),
    );

    return {
      total: trades.total,
      totalTaxFee: trades.totalTaxFee,
      totalMilitaryFee: trades.totalMilitaryFee,
      deals,
    };
  }

  async setDeal(
    purchaseDeal: ITrades,
    sellDeal: ITrades,
    sellComission?: number,
  ): Promise<Deal> {
    const [purchaseRate, saleRate] = await this.fetchPurchaseAndSellRate(
      purchaseDeal,
      sellDeal,
    );

    const quantityToProcess = Math.min(purchaseDeal.q, sellDeal.q);

    const deal = this.getDeal({
      ticker: purchaseDeal.instr_nm,
      purchaseCommission: purchaseDeal.commission,
      purchaseDate: new Date(purchaseDeal.date),
      purchasePrice: purchaseDeal.p,
      purchaseRate,
      quantity: quantityToProcess,
      saleCommission: sellComission * purchaseDeal.q,
      saleDate: new Date(sellDeal.date),
      salePrice: sellDeal.p,
      saleRate,
    });

    sellDeal.q -= quantityToProcess;
    purchaseDeal.q -= quantityToProcess;

    return deal;
  }

  async fetchPurchaseAndSellRate(purchaseDeal: ITrades, sellDeal: ITrades) {
    const [{ rate: purchaseRate }, { rate: sellRate }] = await Promise.all([
      this.currencyExchangeService.getCurrencyExchange(
        purchaseDeal.curr_c,
        purchaseDeal.date,
      ),

      this.currencyExchangeService.getCurrencyExchange(
        sellDeal.curr_c,
        sellDeal.date,
      ),
    ]);

    return [purchaseRate, sellRate];
  }

  findDealByDateAndPrice(deals: ITrades[], currentDeal: ITrades) {
    return deals.find(
      (deal) =>
        this.currencyExchangeService.formatDateForCurrencyExchange(
          deal.date,
        ) ===
          this.currencyExchangeService.formatDateForCurrencyExchange(
            currentDeal.date,
          ) && deal.p === currentDeal.p,
    );
  }

  getTotalTaxFee(total: number) {
    if (total <= 0) {
      return 0;
    }

    return total * 0.18;
  }

  getMilitaryFee(total: number) {
    if (total <= 0) {
      return 0;
    }

    return total * 0.015;
  }

  getDeal({
    ticker,
    quantity,
    purchaseCommission,
    purchaseDate,
    purchasePrice,
    purchaseRate,
    saleCommission,
    saleDate,
    salePrice,
    saleRate,
  }: DealOptions = {}) {
    const purchaseSum = purchasePrice * quantity;
    const _purchaseCommission = purchaseCommission || purchaseSum;
    const saleSum = salePrice * quantity;
    const _saleCommission = saleCommission || saleSum;
    const purchaseUah =
      (purchaseSum + _purchaseCommission) * purchaseRate +
      _saleCommission * saleRate;
    const saleUah = saleSum * saleRate;

    return {
      id: Symbol(),
      purchase: {
        commission: _purchaseCommission,
        date: purchaseDate,
        price: purchasePrice,
        rate: purchaseRate,
        sum: purchaseSum,
        uah: purchaseUah,
      },
      quantity: quantity,
      sale: {
        commission: _saleCommission,
        date: saleDate,
        price: salePrice,
        rate: saleRate,
        sum: saleSum,
        uah: saleUah,
      },
      ticker: ticker,
      total: saleUah - purchaseUah,
      percent: saleUah / purchaseUah - 1,
    };
  }

  async calculateDividends(file: Express.Multer.File) {
    const corporateActions = this.readReport(file).corporate_actions
      .detailed as ICorporateAction[];

    const filteredActionsByDividend = corporateActions.filter(
      (action) => action.type_id === 'dividend',
    );

    const dividends = await Promise.all(
      filteredActionsByDividend.map(async (dividend) => {
        const { rate } = await this.currencyExchangeService.getCurrencyExchange(
          dividend.currency,
          dividend.date,
        );

        return {
          ticker: dividend.ticker,
          rate,
          price: dividend.amount,
          uah: dividend.amount * rate,
        };
      }),
    );

    const totalDividends = dividends.reduce(
      (acc, dividend) => acc + dividend.uah,
      0,
    );

    return {
      total: {
        sumUAH: totalDividends,
        taxFee: totalDividends * 0.09,
        militaryFee: totalDividends * 0.015,
      },
      dividendsResult: dividends,
    };
  }
}
