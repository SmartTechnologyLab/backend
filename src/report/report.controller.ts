import {
  Controller,
  Get,
  Post,
  Query,
  UploadedFiles,
  UseInterceptors,
} from '@nestjs/common';
import { ReportService } from './report.service';
import { FilesInterceptor } from '@nestjs/platform-express';
import { IReport, ITrades } from './types';
import * as fs from 'fs';
import * as path from 'path';
import { sortByDate } from './helpers';
import { reverse } from 'ramda';

@Controller('report')
export class ReportController {
  constructor(private reportService: ReportService) {}

  private filePath21 = path.join(__dirname, '..', '..', '21.json');
  private filePath22 = path.join(__dirname, '..', '..', '22.json');
  private filePath23 = path.join(__dirname, '..', '..', '23.json');

  @Get('test')
  async test() {
    return JSON.parse(fs.readFileSync(this.filePath21, 'utf-8'));
  }

  @Post('deals')
  @UseInterceptors(FilesInterceptor('file', 10))
  async getDeals(
    @Query('type') type: string,
    @UploadedFiles() files: Express.Multer.File[],
  ) {
    try {
      const getReportFunction =
        type === 'extended'
          ? this.reportService.getReportExtended.bind(this.reportService)
          : this.reportService.getReport.bind(this.reportService);

      if (files.length === 1) {
        return getReportFunction(
          this.reportService.readReport(files.at(0)).trades.detailed,
          false,
        ) as Promise<ITrades[]>;
      }

      const reports: IReport[] = [];
      const dealsToCalculate: ITrades[] = [];

      files.forEach((file) => {
        reports.push(this.reportService.readReport(file));
      });

      const sortedReportsByDate = sortByDate(reports);

      const reversedReports: IReport[] = reverse(sortedReportsByDate);

      const remainedDealsMap = new Map<number, ITrades[]>();

      for (const [indx, statement] of Object.entries(reversedReports)) {
        const index = +indx;

        if (index !== sortedReportsByDate.length - 1) {
          const previousDeals = remainedDealsMap.get(index - 1) || [];

          const deals = await getReportFunction(
            previousDeals.length
              ? [...previousDeals, ...statement.trades.detailed]
              : statement.trades.detailed,
            true,
          );

          if (!remainedDealsMap.has(index)) {
            remainedDealsMap.set(index, []);
          }

          if ((deals as ITrades[]).length) {
            remainedDealsMap.get(index)?.push(...(deals as ITrades[]));
          }

          if (
            index === sortedReportsByDate.length - 2 &&
            (deals as ITrades[]).length
          ) {
            dealsToCalculate.push(...(deals as ITrades[]));
          }
        }
      }

      return getReportFunction(
        [...dealsToCalculate, ...sortedReportsByDate.at(0).trades.detailed],
        false,
      ) as Promise<ITrades[]>;
    } catch (error) {
      throw new Error(error);
    }
  }
}
