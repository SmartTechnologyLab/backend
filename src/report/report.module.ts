import { Module } from '@nestjs/common';
import { ReportService } from './report.service';
import { ReportController } from './report.controller';
import { CurrencyExchangeModule } from 'src/currencyExchange/currencyExchange.module';
import { CurrencyExchangeService } from 'src/currencyExchange/currencyExchange.service';
import { NormalizeTradesModule } from 'src/normalizeTrades/normalizeTrades.module';
import { NormalizeTradesService } from 'src/normalizeTrades/normalizeTrades.service';

@Module({
  imports: [CurrencyExchangeModule, NormalizeTradesModule],
  controllers: [ReportController],
  providers: [ReportService, CurrencyExchangeService, NormalizeTradesService],
})
export class ReportModule {}
