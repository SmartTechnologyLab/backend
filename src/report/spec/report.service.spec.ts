import { ReportService } from '../report.service';
import { CurrencyExchangeService } from '../../currencyExchange/currencyExchange.service';
import { trades, tradesNextYear } from './__fixtures__/report';
import { expectedGroupedTrades } from './__fixtures__/dealsExtended';
import { Deal, IDealReport, ITrades } from '../types';

const mockGetCurrencyExchange = jest.fn().mockResolvedValue({ rate: 44 });
const mockFormatDateForCurrencyExchange = jest.fn();

jest.mock('../../currencyExchange/currencyExchange.service', () => {
  return {
    CurrencyExchangeService: jest.fn().mockImplementation(() => {
      return {
        getCurrencyExchange: mockGetCurrencyExchange,
        formatDateForCurrencyExchange: mockFormatDateForCurrencyExchange,
      };
    }),
  };
});

describe('Report Service', () => {
  let currencyExchangeService: CurrencyExchangeService;
  let reportService: ReportService;

  beforeEach(() => {
    currencyExchangeService = new CurrencyExchangeService();
    reportService = new ReportService(currencyExchangeService);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('service should be defined', () => {
    expect(reportService).toBeDefined();
  });

  describe('group trades by ticker', () => {
    it('group trades by ticker', () => {
      const groupedTrades = reportService.groupTradesByTicker(trades);

      expect(groupedTrades).toMatchSnapshot();
    });

    it('ticker not contains extra signs', () => {
      const groupedTrades = reportService.groupTradesByTicker(trades);

      expect(Object.keys(groupedTrades)).toHaveLength(
        Object.keys(expectedGroupedTrades).length,
      );

      Object.entries(groupedTrades).forEach(([, trades]) => {
        trades.forEach((trade) => {
          expect(trade.instr_nm).not.toContain('.');
        });
      });
    });
  });

  describe('report extended', () => {
    it('calculated deals', async () => {
      const TOTAL = -2834.19;
      const TAXES = 0;

      const dealsExtended = (await reportService.getReportExtended(
        trades,
      )) as IDealReport<Deal>;

      expect(dealsExtended.deals).toHaveLength(8);

      expect(+dealsExtended.total.toFixed(2)).toEqual(TOTAL);

      expect(dealsExtended.totalMilitaryFee).toEqual(TAXES);

      expect(dealsExtended.totalMilitaryFee).toEqual(TAXES);
    });

    it('returns object with certain keys', async () => {
      const dealsExtended = (await reportService.getReportExtended(
        trades,
      )) as IDealReport<Deal>;

      expect(Object.keys(dealsExtended).length).toEqual(4);
    });

    it('throw an error when trade with operation sell is left', async () => {
      await expect(async () => {
        await reportService.getReportExtended(tradesNextYear);
      }).rejects.toThrow('Not enough buy deals');
    });

    it('setting isPrevDeal to true returns trades that were not sold', async () => {
      const remainedDeals = (await reportService.getPrevTrades(
        trades,
      )) as ITrades[];

      expect(remainedDeals).toMatchSnapshot();
      expect(remainedDeals).toHaveLength(1);
      expect(Array.isArray(remainedDeals)).toBeTruthy();
    });
  });

  describe('getReport', () => {
    it('returns deals that is summed by ticker', async () => {
      const report = await reportService.getReport(trades);

      expect(Object.keys(report).length).toEqual(4);
      expect(report).toMatchSnapshot();
    });
  });

  describe('getTotalTaxFee', () => {
    it('when total is less than 0 - return 0', () => {
      const TOTAL = -10;

      const taxFee = reportService.getTotalTaxFee(TOTAL);

      expect(taxFee).toStrictEqual(0);
    });

    it('when total is more than 0 - return 0.18% of total', () => {
      const TOTAL = 1000;
      const FEE = 0.18;

      const taxFee = reportService.getTotalTaxFee(1000);

      expect(taxFee).toStrictEqual(TOTAL * FEE);
    });
  });

  describe('getMilitaryFee', () => {
    it('when total is less than 0 - return 0', () => {
      const TOTAL = -10;

      const taxFee = reportService.getMilitaryFee(TOTAL);

      expect(taxFee).toStrictEqual(0);
    });

    it('when total is more than 0 - return 0.015% of total', () => {
      const TOTAL = 1000;
      const FEE = 0.015;

      const taxFee = reportService.getMilitaryFee(1000);

      expect(taxFee).toStrictEqual(TOTAL * FEE);
    });
  });

  describe('find deal by date and price', () => {
    beforeEach(() => {
      mockFormatDateForCurrencyExchange.mockReturnValue('20210607');
    });

    it('founds deal when date and price is equal', () => {
      const trade: ITrades = {
        date: '2021-06-07 11:48:21',
        instr_nm: 'FIPO',
        operation: 'buy',
        p: 35.53,
        curr_c: 'USD',
        q: 13,
        summ: 461.89,
        commission: 2.31,
      };

      const founDeal = reportService?.findDealByDateAndPrice(
        expectedGroupedTrades.FIPO,
        trade,
      );

      expect(founDeal).toEqual(trade);
    });

    it('founds deal when date and price is equal', () => {
      const trade: ITrades = {
        date: '2021-06-07 11:48:21',
        instr_nm: 'FIPO',
        operation: 'buy',
        p: 100.53,
        curr_c: 'USD',
        q: 13,
        summ: 461.89,
        commission: 2.31,
      };

      const foundDeal = reportService?.findDealByDateAndPrice(
        expectedGroupedTrades.FIPO,
        trade,
      );

      expect(foundDeal).not.toEqual(trade);
    });
  });

  describe('fetch purchase and sell rate', () => {
    const purchaseTrade: ITrades = {
      date: '2021-06-07 11:48:21',
      instr_nm: 'FIPO',
      operation: 'buy',
      p: 100.53,
      curr_c: 'USD',
      q: 13,
      summ: 461.89,
      commission: 2.31,
    };

    const sellTrade: ITrades = {
      date: '2021-06-07 11:48:21',
      instr_nm: 'FIPO',
      operation: 'sell',
      p: 100.53,
      curr_c: 'USD',
      q: 13,
      summ: 461.89,
      commission: 2.31,
    };

    it('when params are setted', async () => {
      const RATE = 44;

      const rates = await reportService.fetchPurchaseAndSellRate(
        purchaseTrade,
        sellTrade,
      );

      expect(mockGetCurrencyExchange).toHaveBeenCalledWith(
        purchaseTrade.curr_c,
        purchaseTrade.date,
      );

      expect(mockGetCurrencyExchange).toHaveBeenCalledWith(
        sellTrade.curr_c,
        sellTrade.date,
      );

      expect(mockGetCurrencyExchange).toHaveBeenCalledTimes(2);
      expect(rates).toEqual([RATE, RATE]);
    });
  });
});
