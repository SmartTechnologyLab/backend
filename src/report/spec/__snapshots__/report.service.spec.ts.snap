// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report Service getReport returns deals that is summed by ticker 1`] = `
{
  "deals": [
    {
      "percent": -0.007445928377260325,
      "purchaseUAH": 40951.24,
      "saleUAH": 40646.32,
      "ticker": "FIPO",
      "total": -304.91999999999825,
    },
    {
      "percent": -0.031465987631866144,
      "purchaseUAH": 4838.24,
      "saleUAH": 4686,
      "ticker": "GOOGL",
      "total": -152.23999999999978,
    },
    {
      "percent": -0.0625,
      "purchaseUAH": 36094.66666666667,
      "saleUAH": 34100,
      "ticker": "MNDY_IPO",
      "total": -1994.666666666667,
    },
    {
      "percent": -0.028919654119113125,
      "purchaseUAH": 5851.5599999999995,
      "saleUAH": 5623.2,
      "ticker": "RIDE",
      "total": -228.3599999999999,
    },
    {
      "percent": -0.030674846625766916,
      "purchaseUAH": 5020.400000000001,
      "saleUAH": 4866.400000000001,
      "ticker": "WISH",
      "total": -154,
    },
  ],
  "total": -2834.1866666666647,
  "totalMilitaryFee": 0,
  "totalTaxFee": 0,
}
`;

exports[`Report Service group trades by ticker group trades by ticker 1`] = `
{
  "APPL": [
    {
      "commission": 1.73,
      "curr_c": "USD",
      "date": "2021-06-10 18:41:12",
      "instr_nm": "APPL",
      "operation": "buy",
      "p": 10.65,
      "q": 10,
      "summ": 106.5,
    },
  ],
  "FIPO": [
    {
      "commission": 2.31,
      "curr_c": "USD",
      "date": "2021-06-07 11:48:21",
      "instr_nm": "FIPO",
      "operation": "buy",
      "p": 35.53,
      "q": 13,
      "summ": 461.89,
    },
    {
      "commission": 2.31,
      "curr_c": "USD",
      "date": "2021-06-07 11:48:21",
      "instr_nm": "FIPO",
      "operation": "buy",
      "p": 35.53,
      "q": 13,
      "summ": 461.89,
    },
    {
      "commission": 2.31,
      "curr_c": "USD",
      "date": "2021-06-07 11:48:21",
      "instr_nm": "FIPO",
      "operation": "sell",
      "p": 35.53,
      "q": 26,
      "summ": 461.89,
    },
  ],
  "GOOGL": [
    {
      "commission": 1.73,
      "curr_c": "sell",
      "date": "2021-06-10 18:41:12",
      "instr_nm": "GOOGL",
      "operation": "sell",
      "p": 10.65,
      "q": 10,
      "summ": 106.5,
    },
    {
      "commission": 1.73,
      "curr_c": "sell",
      "date": "2021-06-11 18:41:12",
      "instr_nm": "GOOGL",
      "operation": "buy",
      "p": 10.65,
      "q": 10,
      "summ": 106.5,
    },
  ],
  "MNDY_IPO": [
    {
      "commission": 7.75,
      "curr_c": "USD",
      "date": "2021-06-10 14:06:17",
      "instr_nm": "MNDY_IPO",
      "operation": "buy",
      "p": 155,
      "q": 1,
      "summ": 155,
    },
    {
      "commission": 7.75,
      "curr_c": "USD",
      "date": "2021-07-10 14:06:17",
      "instr_nm": "MNDY_IPO",
      "operation": "buy",
      "p": 156,
      "q": 3,
      "summ": 155,
    },
    {
      "commission": 7.75,
      "curr_c": "sell",
      "date": "2021-10-10 14:06:17",
      "instr_nm": "MNDY_IPO",
      "operation": "sell",
      "p": 155,
      "q": 3,
      "summ": 155,
    },
    {
      "commission": 7.75,
      "curr_c": "sell",
      "date": "2021-11-10 14:06:17",
      "instr_nm": "MNDY_IPO",
      "operation": "sell",
      "p": 155,
      "q": 1,
      "summ": 155,
    },
  ],
  "RIDE": [
    {
      "commission": 1.73,
      "curr_c": "USD",
      "date": "2021-06-10 18:41:12",
      "instr_nm": "RIDE",
      "operation": "buy",
      "p": 10.65,
      "q": 10,
      "summ": 106.5,
    },
    {
      "commission": 1.73,
      "curr_c": "sell",
      "date": "2021-06-10 18:41:12",
      "instr_nm": "RIDE",
      "operation": "sell",
      "p": 10.65,
      "q": 12,
      "summ": 106.5,
    },
    {
      "commission": 1.73,
      "curr_c": "sell",
      "date": "2021-06-10 18:41:12",
      "instr_nm": "RIDE",
      "operation": "buy",
      "p": 10.65,
      "q": 2,
      "summ": 106.5,
    },
  ],
  "WISH": [
    {
      "commission": 1.75,
      "curr_c": "USD",
      "date": "2021-06-10 16:30:00",
      "instr_nm": "WISH",
      "operation": "buy",
      "p": 11.06,
      "q": 10,
      "summ": 110.6,
    },
    {
      "commission": 1.75,
      "curr_c": "USD",
      "date": "2021-06-10 16:30:00",
      "instr_nm": "WISH",
      "operation": "sell",
      "p": 11.06,
      "q": 10,
      "summ": 110.6,
    },
  ],
}
`;

exports[`Report Service report extended setting isPrevDeal to true returns trades that were not sold 1`] = `
[
  {
    "commission": 1.73,
    "curr_c": "USD",
    "date": "2021-06-10 18:41:12",
    "instr_nm": "APPL",
    "operation": "buy",
    "p": 10.65,
    "q": 10,
    "summ": 106.5,
  },
]
`;
